pkg/istio-agent/healthcheck.go

```diff
@@ -0,0 +1,227 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istioagent
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"time"
+
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/genproto/googleapis/rpc/status"
+)
+
+type HealthCheckType string
+
+const (
+	HTTPHealthCheck   HealthCheckType = "HTTP"
+	TCPHealthCheck    HealthCheckType = "TCP"
+	ExecHealthCheck   HealthCheckType = "Exec"
+	HealthInfoTypeURL string          = "type.googleapis.com/istio.v1.HealthInformation"
+)
+
+type WorkloadHealthChecker struct {
+	config ApplicationHealthCheckConfig
+}
+
+type ApplicationHealthCheckConfig struct {
+	InitialDelay   time.Duration
+	ProbeTimeout   time.Duration
+	CheckFrequency time.Duration
+	SuccessThresh  int
+	FailThresh     int
+	CheckType      HealthCheckType
+	HTTPConfig     HTTPHealthCheckConfig
+	TCPConfig      TCPHealthCheckConfig
+	ExecConfig     ExecHealthCheckConfig
+}
+
+type HTTPHealthCheckConfig struct {
+	Path    string
+	Port    uint32
+	Scheme  string
+	Headers map[string][]string
+}
+
+type TCPHealthCheckConfig struct {
+	Host string
+	Port string
+}
+
+type ExecHealthCheckConfig struct {
+	ExecutableName string
+	Args           []string
+}
+
+// PerformApplicationHealthCheck Performs the application-provided configuration health check.
+// Designed to run async.
+// TODO:
+// 	- Add channel param for quit (better error handling as well)
+// 	- Because there are 3 possible configs, there are 3 possible healthcheck paths, and all of them
+// 		are defined here. Therefore, there is quite a bit of duplicate code in success/fail threshold
+// 		and healthChannel sending. This code should be better.
+// 	- Should the CheckFrequency Delay be a time.Ticker?
+func (w *WorkloadHealthChecker) PerformApplicationHealthCheck(notifyHealthChange chan *discovery.DiscoveryRequest) {
+	// delay before starting probes.
+	time.Sleep(w.config.InitialDelay)
+
+	// tracks number of success & failures after last success/failure
+	numSuccess, numFail := 0, 0
+	// if the last send/event was a success, this is true, by default false because we want to
+	// first send a healthy message.
+	lastStateHealthy := false
+	if w.config.CheckType == HTTPHealthCheck {
```

Comment:
Prefer split each type of healthcheck out, and make this entry point more clear

Reply:
How do you think we should split it out?

Right now I have the config and it holds all the values, and the method selects the correct config.

should all of these health check routines be different methods?

Reply:
looks basically the same in k8s: 
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/prober/prober.go#L156